### PR TITLE
[SYCL] Merge two set_final_data methods into one

### PR DESCRIPTION
**Pull request by maximdimakov**
This is a follow-up after intel/llvm#3099. To simplify the code it was decided to merge two set_final_data methods.

### CODE REVIEW

**Review by romanovvlad — COMMENTED** [review]
LGTM, but would like @s-kanaev to approve.

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp [hunk 1]
```diff
@@ -184,23 +178,8 @@ class __SYCL_EXPORT SYCLMemObjT : public SYCLMemObjI {
...
-  }
-
-  template <typename Destination>
-  __SYCL_DLL_LOCAL EnableIfOutputIteratorBool<Destination>
-  set_final_data(Destination FinalData) {
-    MUploadDataFunctor = [this, FinalData]() {
-      using DestinationValueT = iterator_value_type_t<Destination>;
-      // TODO if Destination is ContiguousIterator then don't create
-      // ContiguousStorage. updateHostMemory works only with pointer to
-      // continuous data.
```
Comment (s-kanaev) [inline sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp:201]: I believe this `TODO` is to be preserved as this is a point of improvement and we don't want to miss it.

**Review by s-kanaev — CHANGES_REQUESTED** [review]
The change looks good. Though, a `TODO` is to be preserved.

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp (hunk 1 above)
Comment (maximdimakov) [reply sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp:201]: TODO is in the combined set_final_data method, 177-179 lines

**Review by s-kanaev — APPROVED** [review]

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp (hunk 1 above)
Comment (s-kanaev) [reply sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp:201]: Oh, my bad. Didn't see it first.

**Pull request merged**